llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -5570,10 +5570,35 @@ Instruction *InstCombinerImpl::foldICmpEquality(ICmpInst &I) {
       Z = B;
     }
 
-    if (X) { // Build (X^Y) & Z
-      Op1 = Builder.CreateXor(X, Y);
-      Op1 = Builder.CreateAnd(Op1, Z);
-      return new ICmpInst(Pred, Op1, Constant::getNullValue(Op1->getType()));
+    if (X) {
+      // (X&P2) == (X&-P2)
+      //    -> X u< P2*2
+      // (X&P2) != (X&-P2)
+      //    -> X u>= P2*2
+      // iff P2 is not INT_MIN
```

Comment:
This looks too specific to me. Can't we generalize this to `X&Y` being a high mask? Something like this? https://alive2.llvm.org/ce/z/RfCMU5 (Not sure what the best way to check for high mask is in alive.)

Reply:
> (Not sure what the best way to check for high mask is in alive.)

A high mask is also a negated power-of-2.
See https://alive2.llvm.org/ce/z/43UGar

Reply:
Although we could be slightly more general in that what we really care about is `(X & HiMask) == (X & LoBitsOfHiMask)` -> `X u< RoundUpP2(LoBitsOfHiMask)`. Vice versa as well. Let me make proofs and ill update to be more general.

Reply:
The proofs end up a bit complicated but:
https://alive2.llvm.org/ce/z/jEhpum

Really what we are doing here is proving that `(X & (HiMask ^ LoMask)) == 0` Then `(X & NegP2) == 0` simplifies to `X u< -NegP2`.

Reply:
I don't understand why you need the specific high mask + low mask pattern. Why is my suggestion to check that `X^Y` is a high mask (or as @dtcxzyw pointed out, a negative power of two) and then fold to `<= ~(X^Y)` not sufficient?

This covers the case where X is a high mask and Y is a low mask of the high mask as a special case, but also covers other cases. E.g. `X = 0b1010, Y = 0b0100, X^Y = 0b1110` is also valid for this transform, does not have X or Y being a high mask themselves.

Reply:
Okay its not so complicated. All we really need to do is relax the restrictions on the existing fold if we have the NegP2 case.